### ISSUE TITLE
Kafka Consumer Groups consuming excessive amounts of memory

### ISSUE THREAD

**Issue by clemens** [getsentry/self-hosted#1394]
### Version

22.2.0

### Steps to Reproduce

Nothing in particular to reproduce – we're running a rather simple setup with ~20 or so apps. Last week, we had roughly 11k errors and 1.3m transactions in total.

### Expected Result

We'd expect our systems to be relatively chill due to the reasonable load we're putting on them.

### Actual Result

2 containers seem to be consuming excessive amounts of memory:

![image](https://user-images.githubusercontent.com/3749/159438401-9216195a-0668-4b76-a523-40021fc40acf.png)

The upper line that constantly keeps increasing is `snuba-subscription-consumer-transactions_production` and the lower line that constantly increases (except that one drop at March 11) is `snuba-subscription-consumer-events_production`.

**Comment by ml14tlc** [comment]
Hi, here's the details

1. Ubuntu 20.04 VM with 4 vCPUs and 16 GB RAM. We run sentry via docker compose (docker 20.10.3, compose 1.28.2)
2. Grafana + Prometheus + cAdvisor
3. We started to run sentry on premise about one year ago. The problem with the memory started with the update from 22.11.0 to 22.2.0 on 7th March

To be more precise:
- problems started with the update to 21.11.0, when some consumers showed the behaviour described in the issue [1204](https://github.com/getsentry/self-hosted/issues/1204), i.e., some consumers filled for hours and then they went back to zero. The cycle would repeat 5 or 6 times a day
- we updated to 22.2.0, the problem above disappeared (the consumers are emptied very often), but the two consumers started to use memory without any limit. Their usage grew without end for days, till the VM ran out of RAM
- increasing the RAM from 8 GB to 16 GB didn't help, same pattern
- updating to 22.3.0 didn't help
- since the update to 22.2.0 I see an increased memory usage by the kafka, postgres and click-house containers (but I suppose it's fine, moreover the usage pattern seems healthy, it increases from time to time, but then it falls. In other words, it doesn't grow indefinitely)
- I regularly restart the two consumer containers. It seems that no information are lost (at least judging from the graphs shown in sentry's UI), so my question is: why this linear growth?
- if I could choose, I'd go back to the behaviour we saw till 21.11.0, i.e., growth of the consumers' queues, but it was somewhat limited. Having containers / processes which are so memory hungry makes it difficult to manage the resources
- we're using the compose file provided by the default installation procedure (we only add some labels for monitoring). Please advise in case we have to limit the container resources (i.e., the two consumer containers are then OOM killed)

**Comment by aminvakil** [comment]
We're seeing the same pattern after upgrading to 22.2.0, we use prometheus + cadvisor + grafana to monitor containers and here is dashboards of 30 days ago. Since April 5th we have bumped sentry to 22.2.0 from 22.1.0.

snuba-consumer:
![image](https://user-images.githubusercontent.com/12948692/162607742-01744baf-a871-4832-8185-c273197d6aa2.png)
sunba-outcomes-consumer:
![image](https://user-images.githubusercontent.com/12948692/162607756-ab1be03c-7236-4e4b-9352-e3a14877b021.png)
snuba-sessions-consumer:
![image](https://user-images.githubusercontent.com/12948692/162607765-35c4b65f-d0bd-4b09-957e-2e943b12156e.png)
sunba-subscription-consumer-events:
![image](https://user-images.githubusercontent.com/12948692/162607783-277f2a92-9a54-4779-8255-08d0617822e2.png)
sunba-subscription-consumer-transactions:
![image](https://user-images.githubusercontent.com/12948692/162607793-0e7f81ce-18e7-49ab-9e6b-83d3c9761825.png)

Here is also memory usage of VM which has 32GB memory:
![image](https://user-images.githubusercontent.com/12948692/162607826-ec68a1c0-fa2d-47c9-97ea-0976de90d8c3.png)

**Comment by chadwhitacre** [comment]
Multiple reports here. What additional info do we need to debug, @getsentry/sns? What hypotheses can we generate?

**Comment by chadwhitacre** [comment]
From https://github.com/getsentry/self-hosted/issues/1431:

### Version

22.4.0

### Steps to Reproduce

1. upgraded from 22.1.0 to 22.4.0


### Expected Result

Memory usage remains stable (as it has for months with 22.1.0)

### Actual Result

Immediately after upgrading to Sentry 22.4.0, we started seeing increased memory usage in our Sentry VM. Memory usage had been stable for months with the previous version (22.1.0). There are no other significant processes running on the VM besides Sentry. Restarting clears the excess memory usage but the leak immediately begins again after docker restart.
![image](https://user-images.githubusercontent.com/1498727/165346823-b5488fd7-04be-48a2-9575-85bd4b817aa6.png)

**Comment by evanh** [comment]
We have identified one possible cause of the memory leak, and are testing it now. If it fixes the problem I will update here. PR is here: https://github.com/getsentry/snuba/pull/2649

**Comment by lynnagara** [comment]
^ This fix has been merged. If anyone wants to try running master and report back on whether that improves anything we'd be grateful. Note: we still suspect that may be another potential memory leak affecting this container since we have been able to reproduce it ourselves, but the fix may improve things somewhat.

**Comment by chadwhitacre** [comment]
From @ml14tlc on the Discussion version of this thread:

> Hi all, we updated to the latest version, the memory consumption didn't show any strange pattern in the last hours. The lag of the Kafka consumers look good as well.

Looks like we're good here. 👍

**Comment by aminvakil** [comment]
> From @aminvakil and @ml14tlc on the Discussion version of this thread:
> 
> > > Now that https://github.com/getsentry/self-hosted/releases/tag/22.5.0 has been released, we can try and test if the issue has been fixed.
> > > As we have this problem in our environment ourselves, I'll try upgrading next week, just wanted to see if someone beats me to it :)
> > 
> > 
> > Hi all, we updated to the latest version, the memory consumption didn't show any strange pattern in the last hours. The lag of the Kafka consumers look good as well.
> 
> Looks like we're good here. +1

Yeah, It's been almost three weeks since we bumped to 22.5.0 and it was solved from our side too.
